src/main/java/at/hannibal2/skyhanni/api/SkillAPI.kt

```diff
@@ -0,0 +1,339 @@
+import at.hannibal2.skyhanni.events.InventoryFullyOpenedEvent
+import at.hannibal2.skyhanni.events.SkillOverflowLevelupEvent
+import at.hannibal2.skyhanni.features.misc.skillprogress.SkillProgress
+import at.hannibal2.skyhanni.features.misc.skillprogress.SkillUtil.activeSkill
+import at.hannibal2.skyhanni.features.misc.skillprogress.SkillUtil.calculateLevelXp
+import at.hannibal2.skyhanni.features.misc.skillprogress.SkillUtil.calculateOverFlow
+import at.hannibal2.skyhanni.features.misc.skillprogress.SkillUtil.getLevel
+import at.hannibal2.skyhanni.features.misc.skillprogress.SkillUtil.getSkillInfo
+import at.hannibal2.skyhanni.features.misc.skillprogress.SkillUtil.levelArray
+import at.hannibal2.skyhanni.features.misc.skillprogress.SkillUtil.levelingMap
+import at.hannibal2.skyhanni.features.misc.skillprogress.SkillUtil.xpRequiredForLevel
+import at.hannibal2.skyhanni.utils.ItemUtils.cleanName
+import at.hannibal2.skyhanni.utils.ItemUtils.getLore
+import at.hannibal2.skyhanni.utils.LorenzUtils
+import at.hannibal2.skyhanni.utils.NumberUtil.addSeparators
+import at.hannibal2.skyhanni.utils.NumberUtil.formatNumber
+import at.hannibal2.skyhanni.utils.NumberUtil.romanToDecimalIfNecessary
+import at.hannibal2.skyhanni.utils.OSUtils
+import at.hannibal2.skyhanni.utils.SimpleTimeMark
+import at.hannibal2.skyhanni.utils.StringUtils.removeColor
+import at.hannibal2.skyhanni.utils.TabListData
+import at.hannibal2.skyhanni.utils.repopatterns.RepoPattern
+import com.google.common.base.Splitter
+import com.google.gson.GsonBuilder
+import com.google.gson.reflect.TypeToken
+import io.github.moulberry.notenoughupdates.util.Constants
+import io.github.moulberry.notenoughupdates.util.Utils
+import io.github.moulberry.notenoughupdates.util.XPInformation
+import net.minecraft.init.Blocks
+import net.minecraft.init.Items
+import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
+import java.util.LinkedList
+import java.util.regex.Matcher
+
+object SkillAPI {
+    private val patternGroup = RepoPattern.group("display.skilldisplay")
+    private val SKILL_PATTERN_PERCENT by patternGroup.pattern("skillpaternpercent", "\\+([\\d.,]+) (.+) \\(([\\d.]+)%\\)")
+    private val SKILL_PATTERN by patternGroup.pattern("skillpattern", "\\+([\\d.,]+) (\\w+) \\(([\\d.,]+)/([\\d.,]+)\\)")
+    private val SKILL_PATTERN_MULTIPLIER by patternGroup.pattern("skillpatternmultiplier", "\\+([\\d.,]+) (.+) \\(([\\d.,]+)/([\\d,.]+[kmb])\\)")
+    private val skillTabPattern by patternGroup.pattern("skilltabpattern", "^§e§lSkills: §r§a(?<type>\\w+) (?<level>\\d+): §r§3(?<progress>.+)%\$")
+    private val maxSkillTabPattern by patternGroup.pattern("maxskilltabpattern", "^§e§lSkills: §r§a(?<type>\\w+) (?<level>\\d+): §r§c§lMAX\$")
+    private val SPACE_SPLITTER = Splitter.on("  ").omitEmptyStrings().trimResults()
+
+    var skillXPInfoMap = mutableMapOf<String, SkillXPInfo>()
+    var oldSkillInfoMap = mutableMapOf<String?, SkillInfo?>()
+    val stackMap = mapOf(
```

Comment:
Please make use a enum for this. I don't like to keep track of string values. And when you at it can you resolve the todo in MinionXp for the enum XpType. Since you working on the api :)